[web-animations] word-spacing should support animating between percentage and fixed values

graouts opened the pull request:
#### 2a6e74ada559cc99db27634a06c4c1b298747417
<pre>
[web-animations] word-spacing should support animating between percentage and fixed values
<a href="https://bugs.webkit.org/show_bug.cgi?id=248218">https://bugs.webkit.org/show_bug.cgi?id=248218</a>

Reviewed by Tim Nguyen.

While we support percentage values for word-spacing, we do not support blending between fixed and percentage values,
so we add the &lt;length-percentage&gt; flag to its animation wrapper. And now that we may produce calculated values as a
result, we account for non-fixed values when serializing the computed style for that property.

There remains the issue where we fail to parse calculated values for word-spacing and this causes a few remaining
FAIL results, this will be dealt with in bug 248220.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):

Canonical link: <a href="https://commits.webkit.org/256951@main">https://commits.webkit.org/256951@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/581c259df8557da931ec3edc46ab83305ff2934e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97323 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6587 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106842 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167105 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101291 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6888 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35324 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89733 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103521 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102986 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83952 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32172 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87027 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75099 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/603 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/586 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5386 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2354 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1830 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41114 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
```diff
@@ -274,13 +274,13 @@ PASS word-break: "keep-all" onto "break-all"
 PASS word-spacing: units "em" onto "rem"
 FAIL word-spacing: units "calc" onto "px" assert_equals: The value should be calc(20% + 30px) at 0ms expected "calc(20% + 30px)" but got "20px"
-FAIL word-spacing: calc assert_equals: The value should be calc(30% + 30px) at 0ms expected "calc(30% + 30px)" but got "0px"
+FAIL word-spacing: calc assert_equals: The value should be calc(30% + 30px) at 0ms expected "calc(30% + 30px)" but got "auto"
```
I don't think `auto` is ever a valid value for `word-spacing`. The default value for word-spacing is `normal`.

Can you check why and fix up so we never return `auto`?

graouts replied:
Yes, `auto` is definitely not correct, and there are other tests suffering from the same problem per EWS. Will look into this.

nt1m replied:
The test failures on the GTK bot are a bit worrying: https://ews-build.webkit.org/#/builders/35/builds/32151

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3610,7 +3610,9 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
-        return zoomAdjustedPixelValue(style.fontCascade().wordSpacing(), style);
+        if (style.wordSpacing().isFixed())
+            return zoomAdjustedPixelValue(style.fontCascade().wordSpacing(), style);
+        return cssValuePool.createValue(style.wordSpacing(), style);
```
Why do we use `style.fontCascade().wordSpacing()` in one case and `style.wordSpacing()` in the other?

graouts replied:
In the case where we have a pre-computed (ie. fixed) value, it is stored on the `fontCascade()` and it's the value we must use to account for zooming as well.

nt1m reviewed (approved):
Looks good, but please check my comments before landing.

graouts committed:
[web-animations] word-spacing should support animating between percentage and fixed values
https://bugs.webkit.org/show_bug.cgi?id=248218

Reviewed by Tim Nguyen.

While we support percentage values for word-spacing, we do not support blending between fixed and percentage values,
so we add the <length-percentage> flag to its animation wrapper. And now that we may produce calculated values as a
result, we account for non-fixed values when serializing the computed style for that property.

There remains the issue where we fail to parse calculated values for word-spacing and this causes a few remaining
FAIL results, this will be dealt with in bug 248220.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):

Canonical link: https://commits.webkit.org/256951@main
